### PR TITLE
[Functions] (Feat): Upgradeable

### PR DESCRIPTION
**Pull request by justinkaseman**
### Description
Prepares the Oracle and Billing Registry contracts to be used with [OpenZeppelin's Upgradable plugin](https://docs.openzeppelin.com/upgrades-plugins/1.x/) using a transparent proxy pattern.

### Code Changes
- `constructor`s changed to `initialize` functions, both within the Oracle / BillingRegistry contracts, along with the contracts that they extend
- Add storage slots to allow adding new variables later, gives 49 32byte slots, per [OpenZeppelin's example](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#storage-gaps)
- Add `WithInit` wrapper contracts that will call `initialize` in a constructor (for use within testing)
- Removes `FunctionsOracleFactory` as the proxy contract replaces it. Version numbers can be seen on the proxy contract by looking at `._getInitializedVersion()` or the implementation contract's `.typeAndVersion()`.

### Dependencies
- added `"@openzeppelin/hardhat-upgrades"`
- changed previous version of `"@openzeppelin/contracts-upgradeable"` to be a pinned second dependency, due to conflicting peer dependencies with `hardhat-upgrades`

### CODE REVIEW

File: contracts/src/v0.8/dev/ConfirmedOwnerUpgradeable.sol
Comment (justinkaseman) [inline contracts/src/v0.8/dev/ConfirmedOwnerUpgradeable.sol:19]: Note: same contract as `ConfirmedOwner`, but extends `Initializable` and swaps `constructor` for `initialize` function

File: contracts/src/v0.8/dev/functions/AuthorizedOriginReceiverUpgradeable.sol
Comment (justinkaseman) [inline contracts/src/v0.8/dev/functions/AuthorizedOriginReceiverUpgradeable.sol:31]: Note: same contract as `AuthorizedOriginReceiver`, but extends `Initializable` and swaps `constructor` for `initialize` function

File: contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol
Comment (justinkaseman) [inline contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol:26]: Note:
same contract as `OCR2Base`, but extends `Initializable` and swaps `constructor` for `initialize` function

File: contracts/src/v0.8/tests/FunctionsOracleMigration.sol
Comment (justinkaseman) [inline contracts/src/v0.8/tests/FunctionsOracleMigration.sol:1]: For reviewing purposes the only difference is the fee on line 149 returns 1 instead of 0.

File: contracts/src/v0.8/tests/FunctionsBillingRegistryMigration.sol
Comment (justinkaseman) [inline contracts/src/v0.8/tests/FunctionsBillingRegistryMigration.sol:1]: For reviewing purposes the only difference is the fee on line 287 returns 1 instead of 0.

**Comment by pinebit** [comment]
This is a truly great work. But it will take time to review it.

File: contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol
Comment (KuphJr) [inline contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol:870]: After reading through the docs, it looks like upgrades should be pretty easy using the provided OpenZepplin HardHat plugin, but I want to make sure we do a test upgrade prior to merging this change.

Comment (KuphJr) [reply contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol:870]: Can we schedule a sync w/ the team to do a run-through of the upgrade process?  Ideally, we should all know how to implement a contract hotfix if necessary.

File: contracts/src/v0.8/tests/FunctionsOracleMigration.sol
Comment (KuphJr) [reply contracts/src/v0.8/tests/FunctionsOracleMigration.sol:1]: SGTM.  Out of curiosity, what is the reason for changing this to 1 instead of 0?

File: contracts/src/v0.8/tests/FunctionsOracleMigrationHelper.sol
Comment (KuphJr) [inline contracts/src/v0.8/tests/FunctionsOracleMigrationHelper.sol:2]: Can you explain the purpose of these helper contracts?

File: contracts/src/v0.8/tests/FunctionsOracleUpgradeableHelper.sol
Comment (KuphJr) [inline contracts/src/v0.8/tests/FunctionsOracleUpgradeableHelper.sol:2]: Is this contract just to test if the migration was successful?

**Review by KuphJr — COMMENTED** [review]
Thanks for getting this PR up!  I had a few questions, but my main request is that we all schedule a sync this week to demo a contract hotfix & migration, then we can go ahead & get this merged!

File: contracts/src/v0.8/tests/FunctionsOracleUpgradeableHelper.sol
Comment (justinkaseman) [reply contracts/src/v0.8/tests/FunctionsOracleUpgradeableHelper.sol:2]: Yep, exactly.

File: contracts/src/v0.8/tests/FunctionsOracleMigrationHelper.sol
Comment (justinkaseman) [reply contracts/src/v0.8/tests/FunctionsOracleMigrationHelper.sol:2]: They just make it a little easier to call fulfillment functions within tests by filling some of the inputs within the contract

File: contracts/src/v0.8/tests/FunctionsOracleWithInit.sol
Comment (justinkaseman) [inline contracts/src/v0.8/tests/FunctionsOracleWithInit.sol:1]: Note: this is a wrapper contract for the go tests which don't have the Upgrades plugin
They _could_ call `initialize` themselves, but this makes it cleaner.

File: contracts/src/v0.8/tests/VRFConsumerV2UpgradeableExample.sol
Comment (justinkaseman) [inline contracts/src/v0.8/tests/VRFConsumerV2UpgradeableExample.sol:7]: Switching VRF to use the pinned version of `contracts-upgradeable`

File: contracts/src/v0.8/tests/FunctionsBillingRegistryWithInit.sol
Comment (justinkaseman) [inline contracts/src/v0.8/tests/FunctionsBillingRegistryWithInit.sol:6]: Note: this is a wrapper for tests that aren't testing with the Upgrades plugin (e.g. Go tests)
They _could_ call `initialize` themselves, but this makes it cleaner.

File: contracts/src/v0.8/dev/functions/AuthorizedOriginReceiverUpgradeable.sol
Comment (pinebit) [inline contracts/src/v0.8/dev/functions/AuthorizedOriginReceiverUpgradeable.sol:24]: All storage variables must begin with s_

File: contracts/package.json
Comment (pinebit) [inline contracts/package.json:81]: Can't we use 4.7.3? Is there anything in 4.8.1 that you ultimately need?

File: contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol
Comment (pinebit) [inline contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol:31]: Have you considered this? https://github.com/smartcontractkit/chainlink/blob/develop/contracts/STYLE.md#visibility

File: contracts/src/v0.8/dev/functions/FunctionsOracle.sol
Comment (pinebit) [inline contracts/src/v0.8/dev/functions/FunctionsOracle.sol:46]: Would it be possible for a malicious user to call this from its own contract constructor in between of this contract deployment and a proxy switch?

**Review by pinebit — COMMENTED** [review]
Added my comments for the first round

File: contracts/src/v0.8/dev/ConfirmedOwnerUpgradeable.sol
Comment (justinkaseman) [reply contracts/src/v0.8/dev/ConfirmedOwnerUpgradeable.sol:2]: It's just copied from the original `ConfirmedOwner`, making as minimal of changes as possible.

File: contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol
Comment (justinkaseman) [reply contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol:31]: This was just lifted from VRFv2, but I will change to maintain our best practices.

File: contracts/src/v0.8/dev/functions/FunctionsOracle.sol
Comment (justinkaseman) [reply contracts/src/v0.8/dev/functions/FunctionsOracle.sol:46]: I like your thinking.

It won't be possible. The initialize functions are [only callable once](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/utils/Initializable.sol#L86).
Any contract upgrades that introduce a new `initialize` function are called in within their own deployment and also only callable once.

File: contracts/package.json
Comment (justinkaseman) [reply contracts/package.json:81]: There have been a couple of actual solidity code changes since `4.7.3`, but doesn't seem like anything vital.
See https://github.com/OpenZeppelin/openzeppelin-contracts/commits/v4.8.1/contracts/proxy/utils/Initializable.sol , everything above "Merge v4.7 back into master".

The reason I used `4.8.1` was more to stay in sync with the latest version that `@openzeppelin/hardhat-upgrades` uses as a peer dependency.

I'm open to us pinning to older versions - what is the argument for doing that? Simplicity in the repo dependencies?

File: contracts/src/v0.8/dev/functions/AuthorizedOriginReceiverUpgradeable.sol
Comment (justinkaseman) [reply contracts/src/v0.8/dev/functions/AuthorizedOriginReceiverUpgradeable.sol:24]: I'll update here and on `AuthorizedOriginReceiver`

File: contracts/package.json
Comment (pinebit) [reply contracts/package.json:81]: I would prefer to keep one of these versions for clarity/simplicity. It looks very odd to keep both versions given they are the same major versions. Have you considered removing 4.7.3?

File: contracts/src/v0.8/dev/functions/FunctionsOracle.sol
Comment (pinebit) [reply contracts/src/v0.8/dev/functions/FunctionsOracle.sol:46]: Right, I know it can be called once. Now imagine hackers know our deployer/signer address (easy to guess from Etherscan). They could be tracking all contract deployments initiated by the address and for all newly deployed contracts they can call initialize(). Then your upgrade flow would break. You would have to repeat deployment, etc. This is a sort of DoS attack, as long as an attacker knows our deploying address. Would it make sense to add some modifier like `onlyOwner` to prevent this?

**Comment by pinebit** [comment]
Everything looks good to me now, but let's resolve the ongoing discussion: https://github.com/smartcontractkit/chainlink/pull/8371#discussion_r1099708833

File: contracts/package.json
Comment (justinkaseman) [reply contracts/package.json:81]: That would force a code change within the VRF contracts that are dependent on 4.7.3.

I tried moving both to local `vendor` contracts, but there is something hardhat doesn't like about that when compiling the artifacts.

File: contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol
Comment (alexroan) [inline contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol:6]: We don't technically need this import as it's implicitly imported via `ConfirmedOwnerUpgradeable`. However, as per our guidelines, let's be explicit about every import in every file so we can reason about inheritance more easily:

```
import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```

We have [Chainlink Solhint](https://github.com/smartcontractkit/chainlink-solhint-rules) rules to enforce this, can we apply these here.

File: contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol
Comment (alexroan) [inline contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol:18]: All of these inherited contracts have their own state, if any were reordered in future versions it would cause storage collisions.

File: contracts/src/v0.8/dev/functions/FunctionsOracle.sol
Comment (alexroan) [inline contracts/src/v0.8/dev/functions/FunctionsOracle.sol:14]: Reorder risk again. We must ensure that future upgrades do not reorder these, or if they extend the previous version, must extend like this:

```
contract FunctionOracle2 is FunctionsOracle, SomethingNew { ... }
```

and not:

```
contract FunctionOracle2 is SomethingNew, FunctionsOracle { ... }
```

How can we enforce? 🤔

File: contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol
Comment (alexroan) [inline contracts/src/v0.8/dev/functions/FunctionsBillingRegistry.sol:24]: Same order comments

File: contracts/src/v0.8/dev/functions/FunctionsOracle.sol
Comment (justinkaseman) [reply contracts/src/v0.8/dev/functions/FunctionsOracle.sol:14]: I believe the Hardhat Upgrades plugin enforces storage clashes, even through dependencies, but lemme run a few quick tests to be sure.

**Review by pinebit — DISMISSED** [review]
From my end all good. Let's address @alexroan feedback and move on.

**Review by CL-Andrew — DISMISSED** [review]

**Review by bolekk — APPROVED** [review]
Re-approving after rebase based on Andrew's earlier approval.

**Comment by PoCk3T-SPAI** [comment]
@justinkaseman @bolekk  is this feature now available? 

I'm trying to make an HTTP GET request using Any API  in my OpenZepellin upgradable smart contract

**Pull request merged**
